File: src/game/adventure/gameobjects/Player.java
package game.adventure.gameobjects;

import game.adventure.util.Constants;

import java.util.ArrayList;
import java.util.List;

public class Player {
    private Item equip;
    private int health;
    private final List<Item> inventory;

    public Player(Builder builder) {
        this.health = builder.health;
        this.inventory = builder.inventory;
        this.equip = builder.equip;
    }

    public boolean isEquip(Item item) {
        return equip.equals(item);
    }

    public void setEquip(String item) {
        setEquip(getItem(item));
    }

    public void removeEquip() {
        setEquip(Constants.GENERIC_ITEM);
    }

    public void addItem(Item newItem) {
        System.out.println("New Item: " + newItem.getName());
        inventory.add(newItem);
    }

    public int getDamage() {
        return getEquip().getDamage();
    }

    public Item getItem(int i) {
        return inventory.get(i);
    }

    public Item getItem(String itemName) {
        for (Item item : inventory) {
            if (item.getName().equalsIgnoreCase(itemName)) return item;
        }
        return Constants.GENERIC_ITEM;
    }

    public boolean hasItem(String itemName) {
        for (Item item : inventory) {
            if (item.getName().equalsIgnoreCase(itemName)) return true;
        }
        return false;
    }

    public void removeItem(Item oldItem) {
        if (isEquip(oldItem)) removeEquip();
        else inventory.remove(oldItem);
    }

    public void removeItem(String oldItem) {
        removeItem(getItem(oldItem));
    }

    public void clearInventory() {
        inventory.clear();
    }

    public List<Item> getInventory() {
        return inventory;
    }

    public Item getEquip() {
        return equip;
    }

    public void setEquip(Item equip) {
        if (!this.equip.equals(Constants.GENERIC_ITEM)) inventory.add(this.equip);
        inventory.remove(equip);
        this.equip = equip;
    }

    public boolean hasEquip() {
        return !getEquip().equals(Constants.GENERIC_ITEM);
    }

    public int getHealth() {
        return health;
    }

    public void setHealth(int health) {
        this.health = health;
    }

    public static class Builder {
        private Item equip = Constants.GENERIC_ITEM;
        private int health = 1;
        private List<Item> inventory = new ArrayList<>();

        public Builder equip(Item equip) {
            this.equip = equip;
            return this;
        }

        public Builder health(int health) {
            this.health = health;
            return this;
        }

        public Builder inventory(List<Item> inventory) {
            this.inventory = inventory;
            return this;
        }

        public Player build() {
            return new Player(this);
        }
    }
}

File: src/game/adventure/gameobjects/Player.kt
package game.adventure.gameobjects

class Player(var health: Int, private val playerInventory: MutableList<Item>, equip: Int) {
    private var equip = 0

    /** equip should be an index in the players inventory. if the number is out of bounds, the player will automatically have no equip. -1 == no equip  */
    init {
        if (equip + 1 > playerInventory.size || equip + 1 < 0 || playerInventory.isEmpty()) {
            this.equip = -1
        } else {
            this.equip = equip
        }
    }

    fun isEquip(item: String): Boolean {
        val index = getInt(item)
        return if (index == -1) {
            false
        } else index == equip
    }

    @Throws(IndexOutOfBoundsException::class)
    fun setEquip(item: String) {
        val index = getInt(item)
        if (index == -1) {
            throw IndexOutOfBoundsException()
        }
        setEquip(index)
    }

    fun removeEquip() {
        setEquip(-1)
    }

    fun addItem(newItem: Item) {
        playerInventory.add(newItem)
    }

    private fun getInt(item: String): Int {
        for (i in playerInventory.indices) {
            if (playerInventory[i].name.equals(item, ignoreCase = true)) {
                return i
            }
        }
        return -1
    }

    val damage: Int
        get() = if (equip == -1) -1 else getItem(equip).damage
    val equipItem: Item
        get() = getItem(equip)

    fun getItem(i: Int): Item {
        return playerInventory[i]
    }

    @Throws(IndexOutOfBoundsException::class)
    fun getItem(item: String): Item {
        val i = getInt(item)
        if (i == -1) {
            throw IndexOutOfBoundsException()
        }
        return playerInventory[i]
    }

    fun hasItem(item: String): Boolean {
        return getInt(item) != -1
    }

    fun removeItem(oldItem: Item) {
        playerInventory.remove(oldItem)
    }

    @Throws(IndexOutOfBoundsException::class)
    fun removeItem(oldItem: String) {
        val index = getInt(oldItem)
        if (index == -1) {
            throw IndexOutOfBoundsException()
        }
        removeItem(getItem(index))
    }

    fun clearInventory() {
        playerInventory.clear()
    }

    val inventory: List<Item>
        get() = playerInventory

    fun setEquip(equip: Int) {
        this.equip = equip
    }

    fun getEquip(): Int {
        return equip
    }
}